netcreateorg/netcreate-2018
# Spring Feature: GroupID and SubGroupID Logging Enhancements

This pull request closes #40 "GroupID Subgroups, Logging". It merges Ben's changes in #43 

**SUBGROUP IDS**
GroupID Access tokens (e.g. `MEEP-ONE-9V3`) can now have **subgroups** of the form `ID#`, where # is an integer. The intended use case is that a group of students will receive their access token, and can then add subgroup IDs to the end of it to identify themselves anonymously. 

Examples of valid tokens now accepted:

* **group id, no subid** `MEEP-ONE-9V3`
* **group id, sub id 1** `MEEP-ONE-9V3-ID1`
* **group id, sub id 291** `MEEP-ONE-9V3-ID291`

Students can pick any number, and then let their teacher know which number they are.

**SUBGROUP ID LOGGING**
Actions logged by the NetCreate server **include the access token**, and this also includes the subgroup ID. Researchers should be able to tell who made what contribution from the logs written to `runtime/logs`. For example:
```
19:27:35 UADDR_05 [MEEP-ONE-9PK-ID2121]	select node	7	John John 17
19:27:47 UADDR_06 [MEEP-ONE-9V3-ID1]	select node	7	John John 17
19:27:53 UADDR_06 [MEEP-ONE-9V3-ID1]	update node	7	{"label":"John John 17","id":7,"attributes":{"Node_Type":"Person","Extra Info":"","Notes":"TEFUCK!!! BONKERS"}}
```
Additionally, the nodes and edges in the database **record the updating client**. While this data isn't currently updated in realtime back to the UI, it is stored in the loki database and can be used to see who created the node and who modified it. 

## TESTING

STARTING TEST PROCEDURE
* open two terminal windows and cd to the `build` directory
* in terminal window 1, run NetCreate with `npm run dev`
* after NetCreate is running, issue the command `npm run logtail` in window 2

TEST LOGIN SUBGROUPIDS
* open Chrome and show the Javascript Console (CMD-OPT-J)
* access NetCreate in read-only mode (i.e. without a valid token)
* enter valid and invalid tokens and try to break it
* add the `-ID#` (where `#` is any positive integer) to the token to see how it changes
* look at Terminal 1 and Terminal 2 as you add/modify/delete nodes and edges

## KNOWN ISSUES
* There is no UI to inspect the node and edge access logs from the client at this time. To confirm operation, Terminal 1 will show status of the `_nlog` and `_elog` properties that are being updated. 
* Updates to these logs are also not currently updated after the initial database load as part of the `SOURCE_UPDATE` message sent to clients.
